src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java

```diff
@@ -0,0 +1,593 @@
+
+        void setProtectedReversePathId(PathId protectedReversePathId);
+
+        Set<PathId> getPathIds();
+
+        Collection<HaFlowPath> getPaths();
+
+        Optional<HaFlowPath> getPath(PathId pathId);
+
+        boolean hasPath(HaFlowPath path);
+
+        void addPaths(HaFlowPath... paths);
+
+        Optional<HaSubFlow> getSubFlow(String subFlowId);
+
+        Set<HaSubFlow> getSubFlows();
+
+        void setSubFlows(Set<HaSubFlow> subFlows);
+
+        FlowStatus getStatus();
+
+        void setStatus(FlowStatus status);
+
+        Instant getTimeCreate();
+
+        void setTimeCreate(Instant timeCreate);
+
+        Instant getTimeModify();
+
+        void setTimeModify(Instant timeModify);
+    }
+
+    /**
+     * POJO implementation of HaFlowData.
+     */
+    @Data
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    static final class HaFlowDataImpl implements HaFlowData, Serializable {
+        private static final long serialVersionUID = 1L;
+        @NonNull String haFlowId;
+        @NonNull HaSharedEndpoint sharedEndpoint;
+        PathId forwardPathId;
+        PathId reversePathId;
+        PathId protectedForwardPathId;
+        PathId protectedReversePathId;
+        boolean allocateProtectedPath;
+        long maximumBandwidth;
+        boolean ignoreBandwidth;
+        boolean strictBandwidth;
+        String description;
+        boolean periodicPings;
+        FlowEncapsulationType encapsulationType;
+        FlowStatus status;
+        Long maxLatency;
+        Long maxLatencyTier2;
+        Instant timeCreate;
+        Instant timeModify;
+        boolean pinned;
+        Integer priority;
+        PathComputationStrategy pathComputationStrategy;
+
+        @Builder.Default
+        @ToString.Exclude
+        @EqualsAndHashCode.Exclude
+        Set<HaSubFlow> subFlows = new HashSet<>();
+
+        @Builder.Default
+        @ToString.Exclude
+        @EqualsAndHashCode.Exclude
+        Set<HaFlowPath> paths = new HashSet<>();
+
+        // The reference is used to link sub-flows back to the ha-flow. See {@link #setSubFlows(Set)}.
+        @Setter(AccessLevel.NONE)
+        @Getter(AccessLevel.NONE)
+        @ToString.Exclude
+        @EqualsAndHashCode.Exclude
+        HaFlow haFlow;
+
+        @Override
+        public Optional<HaSubFlow> getSubFlow(String subFlowId) {
+            return subFlows.stream()
+                    .filter(subFlow -> subFlow.getHaSubFlowId().equals(subFlowId))
+                    .findAny();
+        }
+
+        @Override
+        public Set<HaSubFlow> getSubFlows() {
+            return Collections.unmodifiableSet(subFlows);
+        }
+
+        @Override
+        public void setSubFlows(Set<HaSubFlow> subFlows) {
+            for (HaSubFlow subFlow : this.subFlows) {
+                boolean keepSubFlow = subFlows.stream()
+                        .anyMatch(n -> n == subFlow || n.getHaSubFlowId().equals(subFlow.getHaSubFlowId()));
```

Comment:
What's the meaning of the first condition? `==` is to check whether it is the same object? But then the next condition is true, except when subFlowId is null. I mean, is the first one really needed here?